eclipse/src/saros/editor/internal/ContributionAnnotationManager.java

```diff
@@ -228,88 +253,112 @@ public void refreshAnnotations(IAnnotationModel model) {
+  private History getHistory(final User source) {
+
+    History result = sourceToHistory.get(source);
+
     if (result == null) {
-      result = new LinkedList<ContributionAnnotation>();
+      result = new History();
       sourceToHistory.put(source, result);
     }
+
     return result;
   }
 
   /**
    * Add a contribution annotation to the annotation model and store it into the history of the
    * associated user. Old entries are removed from the history and the annotation model.
    */
-  private void addContributionAnnotation(ContributionAnnotation annotation, Position position) {
+  private void addContributionAnnotation(
+      final ContributionAnnotation annotation, final Position position) {
 
-    annotation.getModel().addAnnotation(annotation, position);
+    final History history = getHistory(annotation.getSource());
+
+    final List<HistoryEntry> removedEntries = removeHistoryEntries(history);
 
-    Queue<ContributionAnnotation> history = getHistory(annotation.getSource());
-    history.add(annotation);
-    while (history.size() > MAX_HISTORY_LENGTH) {
-      ContributionAnnotation oldAnnotation = history.remove();
-      oldAnnotation.getModel().removeAnnotation(oldAnnotation);
+    final HistoryEntry newEntry = new HistoryEntry(annotation, history.currentInsertStamp);
+
+    history.currentInsertStamp = (history.currentInsertStamp + 1) % MAX_HISTORY_LENGTH;
+
+    history.entries.add(newEntry);
+
+    if (removedEntries.isEmpty()) {
+      annotation.getModel().addAnnotation(annotation, position);
+      return;
     }
+
+    // TODO OPTIMIZE it is usually 2 annotations after the queue is full
+    annotation.getModel().addAnnotation(annotation, position);
+
+    for (final HistoryEntry entry : removedEntries)
+      entry.annotation.getModel().removeAnnotation(entry.annotation);
   }
 
   /**
-   * Removes an annotation from the user's history and the annotation model.
-   *
-   * @param annotation
+   * Removes entries from the history based on the history current insert stamp. This method <b>DOES
+   * NOT</b> alter the annotation model the removed annotations in the history belong to!
    */
-  private void removeFromHistory(ContributionAnnotation annotation) {
-    getHistory(annotation.getSource()).remove(annotation);
-    annotation.getModel().removeAnnotation(annotation);
+  private List<HistoryEntry> removeHistoryEntries(final History history) {
+
+    final List<HistoryEntry> removedEntries = new ArrayList<>();
+
+    final int insertStampToRemove = history.currentInsertStamp;
+
+    /* the logic assumes that the entry order does not change during lifetime
+     * i.e if we have a history of size 4 the list must look like this regarding the insert stamps
+     * 0 0 0 1 2 3 3 3 0 1 1 1 2 2 2 2 3 0 1 1 1 1 1 2 3 ... and so on
+     *
+     */
+
+    final Iterator<HistoryEntry> it = history.entries.iterator();
+
+    if (!it.hasNext()) return removedEntries;
```

Comment:
So `if history.entries.isEmpty()`? Would be more readable in my opinion. (Maybe even add an `isEmpty` function to the history class.)